docker-base-build/install_pinned.py

```diff
@@ -0,0 +1,377 @@
+    and the remaining non-blank lines are passed through
+    :func:`parse_requirement`.
+    """
+    if _is_url(filename):
+        with urllib.request.urlopen(filename) as raw:
+            # This is quick-n-dirty; using a proper library like requests
+            # would determine the charset from the content type.
+            with codecs.getreader('utf-8')(raw) as enc:
+                yield from _parse_requirements(enc, filename, constraint=constraint, weak=weak)
+    else:
+        with open(filename) as f:
+            yield from _parse_requirements(f, filename, constraint=constraint, weak=weak)
+
+
+def merge_packages(pkg1: Package, pkg2: Package) -> Package:
+    """Combine two requirements for the same package."""
+    # Simplify some of the logic by ensuring that if one is weak and one is
+    # strong, the strong one is first.
+    if pkg1.weak and not pkg2.weak:
+        pkg1, pkg2 = pkg2, pkg1
+
+    req1 = pkg1.requirement
+    req2 = pkg2.requirement
+    if req1.name != req2.name:
+        raise ValueError(f'Cannot merge requirements with different names: {req1} vs {req2}')
+    # If either requirement is inapplicable because of a marker, just ignore it.
+    if req1.marker and not req1.marker.evaluate():
+        return pkg2
+    if req2.marker and not req2.marker.evaluate():
+        return pkg1
+    if req1.url is not None and req2.url is not None and req1.url != req2.url:
+        if pkg1.weak == pkg2.weak:
+            raise ValueError(f'Cannot merge requirements with inconsistent URLs: {req1} vs {req2}')
+    new_req = Requirement(str(req1))
+    new_req.extras |= req2.extras
+    # Strong specifier only replaces weak if it has an exact version pinned
+    if pkg1.weak == pkg2.weak or not has_exact(new_req.specifier):
+        new_req.specifier &= req2.specifier
+    if req1.url is None and req2.url is not None:
+        new_req.url = req2.url
+    # Cannot have both URL and specifiers. Assume URLs represent some
+    # infinitely high version number, so that they always satisfy >=
+    # specifiers but not others.
+    if new_req.url is not None:
+        for spec in new_req.specifier:
+            if '999999999' not in spec:
+                raise ValueError(f'Cannot combine URL {new_req.url!r} with specifier {spec}')
+        new_req.specifier = SpecifierSet('')
+    return Package(new_req, constraint=pkg1.constraint and pkg2.constraint, weak=pkg1.weak)
+
+
+def version_from_requirement(requirement: Requirement) -> str:
+    if requirement.url is not None:
+        raise ValueError('Cannot get version from an URL requirement')
+    pin = None
+    for spec in requirement.specifier:
+        if spec.operator in {'==', '==='} and '*' not in spec.version:
+            pin = spec.version
+    if pin is None:
+        raise ValueError(f'No version pinned for {requirement.name}')
+    for spec in requirement.specifier:
+        if pin not in spec:
```

Comment:
Is this a magic range check as opposed to a substring check? :-)

Reply:
Yes, `Specifier` implements `__contains__` to check if a version string meets the specification.